### DISCUSSION TITLE
vips_image_write causing slow down when multi-threaded

### DISCUSSION THREAD

**Discussion by Bytologic** [libvips/libvips#2881] [Q&A] [answered]
Hi,

I am trying to extract the data from a VipsImage to my own buffer. To do this, I have created my own function based on `vips_image_write_to_memory`. The function currently looks like this:
```
std::shared_ptr<MemoryPoolLocation<unsigned char>> VipsImageToDataLocation(
    VipsImage* in,
    const std::shared_ptr<ResizableMemoryPool<unsigned char>>& memoryPool)
{
    const auto size = VIPS_IMAGE_SIZEOF_IMAGE(in);
    auto dataLocation = memoryPool->Allocate(size);

    auto tempImage = vips_image_new_from_memory(
        dataLocation->dataPointer,
        size,
        in->Xsize,
        in->Ysize,
        in->Bands,
        in->BandFmt);

    if (vips_image_write(in, tempImage))
    {
        g_object_unref(tempImage);
        return(NULL);
    }
    g_object_unref(tempImage);

    return dataLocation;
}
```

It uses my own implemented memory pool to allocate memory, `dataLocation->dataPointer` points to this buffer.

What I am seeing is that when this function is called from a lot (100+) of threads at the same time, it will occasionally slow down the whole process by almost 10 times. When I look in my Task Manager, I see that the amount of threads that are running on the system fluctuates between 5700 and 13500 when this happens.
I have isolated this problem to `vips_image_write(in, tempImage)`. The slowdown will not happen when I comment that line.

I have tried the following alternatives:

- Use `vips_image_write_to_memory` and copy the resulting buffer into my own buffer. This gave me the same problem.
- Use `vips_image_get_data` and copy the resulting buffer into my own buffer. Gave warning: _VIPS-WARNING **: 14:10:54.087: rewinding image with active region_ and eventually crashed.
- Added std::counting_semaphore<100> to limit the amount of function calls at the same time. This slightly helped since the amount of threads would only fluctuate between 5700 and 6900 and produce less of a slow down, but it slowed the process as a whole slightly.

Do you have any idea what is happening here and how I can solve it?

**Comment by jcupitt** [comment, marked as answer]
Hi @Bytologic,

libvips is a lazy library, so a VipsImage isn't a data structure like an array, it's a function call that can generate pixels on demand. It represents the end of a pipeline, not an array of computed values.

When you call `vips_image_write()`, it will allocate a threadpool and use the function to compute all the pixels in the image. By default, libvips will make a pool as large as the number of hardware threads on your computer. This can be pretty big, of course.

If you expect to be running many `vips_image_write()` at the same time, it might be a good idea to limit the threadpool size, eg.:

```C
vips_concurrency_set( 4 );
```

To use 4 threads for each evaluation.

**Reply by Bytologic** [reply]
Not only did this fix the slow down issue, setting the concurrency to 1 improved the performance by almost 100% when lots of threads are involved! Thank you very much!

**Reply by jcupitt** [reply]
Ah, glad it helped! You'll find memory use is down too.

Yes, there's no point in using a large threadpool if there's not much parallelism in your pipeline.
